File: iOSReactNativeSample/ios/iOSReactNativeSample/BannerAdView.m
#import <React/RCTViewManager.h>
#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>
#import <AdWhaleSDK/AdWhaleSDK.h>

@interface BannerAdViewContainer : UIView <AdWhaleBannerDelegate>
@property (nonatomic, strong) AdWhaleBannerAd *bannerView;
@property (nonatomic, copy) RCTBubblingEventBlock onAdLoaded;
@property (nonatomic, copy) RCTBubblingEventBlock onAdLoadFailed;
@property (nonatomic, copy) RCTBubblingEventBlock onAdClicked;
@property (nonatomic, copy) NSString *adSize;
@property (nonatomic, copy) NSString *placementUid;
@property (nonatomic, assign) BOOL shouldLoadAd;
@end

@implementation BannerAdViewContainer

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _adSize = @"BANNER320x50";
        _placementUid = @"";
        _shouldLoadAd = NO;
    }
    return self;
}

- (void)setAdSize:(NSString *)adSize {
    _adSize = adSize;
    [self updateAdSize];
}

- (void)setPlacementUid:(NSString *)placementUid {
    _placementUid = placementUid;
    [self.bannerView setAdUnitID:placementUid];
}

- (void)setShouldLoadAd:(BOOL)shouldLoadAd {
    _shouldLoadAd = shouldLoadAd;
    if (shouldLoadAd) {
        [self loadAd];
    }
}

- (void)updateAdSize {
    if (!self.bannerView) return;
    CGFloat width = 320;
    CGFloat height = 50;
    if ([_adSize isEqualToString:@"BANNER320x100"]) height = 100;
    else if ([_adSize isEqualToString:@"BANNER300x250"]) height = 250;
    CGRect frame = CGRectMake((UIScreen.mainScreen.bounds.size.width - width) / 2, 0, width, height);
    self.bannerView.frame = frame;
    if ([_adSize isEqualToString:@"BANNER320x50"]) [self.bannerView setAdSize:AdWhaleAdSizeBanner];
    else if ([_adSize isEqualToString:@"BANNER320x100"]) [self.bannerView setAdSize:AdWhaleAdSizeLargeBanner];
    else if ([_adSize isEqualToString:@"BANNER300x250"]) [self.bannerView setAdSize:AdWhaleAdSizeMediumRectangle];
}

- (void)loadAd {
    if (!self.bannerView) {
        CGFloat width = 320;
        CGFloat height = 50;
        if ([_adSize isEqualToString:@"BANNER320x100"]) height = 100;
        else if ([_adSize isEqualToString:@"BANNER300x250"]) height = 250;
        CGRect frame = CGRectMake((UIScreen.mainScreen.bounds.size.width - width) / 2, 0, width, height);
        self.bannerView = [[AdWhaleBannerAd alloc] initWithFrame:frame];
        [self addSubview:self.bannerView];
        [self updateAdSize];
        [self.bannerView setAdUnitID:self.placementUid ?: @""];
        [self.bannerView setRootViewController:RCTPresentedViewController()];
        [self.bannerView setDelegate:self];
    }
    [self.bannerView load];
}

// MARK: - AdWhaleBannerDelegate
- (void)bannerViewDidReceiveAd:(AdWhaleBannerAd *)bannerView {
    if (self.onAdLoaded) self.onAdLoaded(@{});
}
- (void)bannerView:(AdWhaleBannerAd *)bannerView didFailToReceiveAdWithError:(NSError *)error {
    if (self.onAdLoadFailed) self.onAdLoadFailed(@{@"statusCode": @0, @"message": error.localizedDescription ?: @""});
}
- (void)bannerViewDidRecordImpression:(AdWhaleBannerAd *)bannerView {}
- (void)bannerViewWillPresentScreen:(AdWhaleBannerAd *)bannerView {
    if (self.onAdClicked) self.onAdClicked(@{});
}
- (void)bannerViewWillDismissScreen:(AdWhaleBannerAd *)bannerView {}
- (void)bannerViewDidDismissScreen:(AdWhaleBannerAd *)bannerView {}

@end

@interface BannerAdView : RCTViewManager
@end

@implementation BannerAdView
RCT_EXPORT_MODULE()

RCT_EXPORT_VIEW_PROPERTY(onAdLoaded, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdLoadFailed, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdClicked, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(adSize, NSString)
RCT_EXPORT_VIEW_PROPERTY(placementUid, NSString)
RCT_EXPORT_VIEW_PROPERTY(shouldLoadAd, BOOL)

- (UIView *)view {
    return [[BannerAdViewContainer alloc] init];
}
@end 

File: iOSExpoSample/ios/iOSExpoSample/BannerAdView.swift
import UIKit
import AdWhaleSDK
import React

@objc(BannerAdView)
class BannerAdView: RCTViewManager {
    override func view() -> UIView! {
        return BannerAdViewContainer()
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
}

@objc(BannerAdViewContainer)
class BannerAdViewContainer: UIView {
    private var bannerView: AdWhaleBannerAd?
    @objc var adSize: String = "BANNER320x50" {
        didSet {
            updateAdSize()
        }
    }
    @objc var placementUid: String = "" {
        didSet {
            bannerView?.setAdUnitID(placementUid)
        }
    }
    @objc var shouldLoadAd: Bool = false {
        didSet {
            if shouldLoadAd {
                loadAd()
            }
        }
    }
    
    @objc var onAdLoaded: RCTDirectEventBlock?
    @objc var onAdLoadFailed: RCTDirectEventBlock?
    @objc var onAdClicked: RCTDirectEventBlock?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupBannerAd()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupBannerAd()
    }
    
    private func setupBannerAd() {
        let screenWidth = UIScreen.main.bounds.width
        let bannerWidth: CGFloat = 320
        let xPosition = (screenWidth - bannerWidth) / 2
        
        let frame = CGRect(x: xPosition,
                          y: 0,
                          width: bannerWidth,
                          height: getBannerHeight(for: adSize))
        
        bannerView = AdWhaleBannerAd(frame: frame)
        if let bannerView = bannerView {
            addSubview(bannerView)
            updateAdSize()
            bannerView.setAdUnitID(placementUid)
            
            if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
               let rootViewController = windowScene.windows.first?.rootViewController {
                bannerView.setRootViewController(rootViewController)
            }
            
            bannerView.setDelegate(self)
        }
    }
    
    private func getBannerHeight(for size: String) -> CGFloat {
        switch size {
        case "BANNER320x50":
            return 50
        case "BANNER320x100":
            return 100
        case "BANNER300x250":
            return 250
        default:
            return 50
        }
    }
    
    private func updateAdSize() {
        guard let bannerView = bannerView else { return }
        
        // 배너 뷰의 크기를 새로운 사이즈에 맞게 업데이트
        let screenWidth = UIScreen.main.bounds.width
        let bannerWidth: CGFloat = 320
        let xPosition = (screenWidth - bannerWidth) / 2
        let newHeight = getBannerHeight(for: adSize)
        
        bannerView.frame = CGRect(x: xPosition,
                                y: bannerView.frame.origin.y,
                                width: bannerWidth,
                                height: newHeight)
        
        switch adSize {
        case "BANNER320x50":
            bannerView.setAdSize(.banner)
            break;
        case "BANNER320x100":
            bannerView.setAdSize(.largeBanner)
            break;
        case "BANNER300x250":
            bannerView.setAdSize(.mediumRectangle)
            break
        default:
            bannerView.setAdSize(.banner)
        }
    }
    
    private func loadAd() {
        bannerView?.setAdUnitID(placementUid)
        bannerView?.load()
    }
}

// MARK: - BannerAd Delegate
extension BannerAdViewContainer: AdWhaleBannerDelegate {
    func bannerViewDidReceiveAd(_ bannerView: AdWhaleSDK.AdWhaleBannerAd) {
        print("[adwhale] [sample Native] 배너 광고 로드 성공")
        onAdLoaded?([:])
    }
    
    func bannerView(_ bannerView: AdWhaleSDK.AdWhaleBannerAd, didFailToReceiveAdWithError error: Error) {
        print("[adwhale] [sample Native] 배너 광고 로드 실패: \(error.localizedDescription)")
        onAdLoadFailed?([
            "statusCode": 0,
            "message": error.localizedDescription
        ])
    }
    
    func bannerViewDidRecordImpression(_ bannerView: AdWhaleSDK.AdWhaleBannerAd) {
        print("[adwhale] [sample Native] 배너 광고 노출")
    }
    
    func bannerViewWillPresentScreen(_ bannerView: AdWhaleSDK.AdWhaleBannerAd) {
        print("[adwhale] [sample Native] 배너 광고 클릭")
        onAdClicked?([:])
    }
    
    func bannerViewWillDismissScreen(_ bannerView: AdWhaleSDK.AdWhaleBannerAd) {
        print("[adwhale] [sample Native] 배너 광고 화면 닫힘 예정")
    }
    
    func bannerViewDidDismissScreen(_ bannerView: AdWhaleSDK.AdWhaleBannerAd) {
        print("[adwhale] [sample Native] 배너 광고 화면 닫힘")
    }
} 
